libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt

```diff
@@ -414,7 +423,9 @@ internal class MapboxTripSession(
             launch {
                 updateEnhancedLocation(status.enhancedLocation, status.keyPoints)
                 updateRouteProgress(status.routeProgress)
-                isOffRoute = status.offRoute
+                mutex.withLock {
```

Comment:
Wouldn't moving this lock a level or 2 up also resolve https://github.com/mapbox/mapbox-navigation-android/issues/3383?

It can be addressed in a follow-up, just something to think about.

Reply:
Yeah when I saw recent https://github.com/mapbox/mapbox-navigation-android/issues/3383 was wondering the same. Will follow up in a separate PR 👍 Thanks for the heads up!

Reply:
👍  haven't looked into kotlin's Mutex much until now, but it is pretty cool because it is a suspending function. Meaning, you can call it from the main thread and it won't block it. Coroutines can be so neat!

https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.sync/-mutex/lock.html
"In a blocking world you'd typically use synchronized or ReentrantLock for that. Coroutine's alternative is called Mutex. It has lock and unlock functions to delimit a critical section. The key difference is that Mutex.lock() is a suspending function. It does not block a thread."

Reply:
But +1 to what lukas is saying, moving it up a level will likely resolve #3383. Curious how long it takes to complete though, critical sections should be lock minimal computation